test/framework/clusterctl/client.go

```diff
@@ -159,19 +165,46 @@ func Upgrade(ctx context.Context, input UpgradeInput) {
 		input.ClusterctlConfigPath = outputPath
 	}
 
-	log.Logf("clusterctl upgrade apply --contract %s --config %s --kubeconfig %s",
-		input.Contract,
-		input.ClusterctlConfigPath,
-		input.KubeconfigPath,
-	)
+	// Check if the user want a custom upgrade
+	isCustomUpgrade := input.CoreProvider != "" ||
+		len(input.BootstrapProviders) > 0 ||
+		len(input.ControlPlaneProviders) > 0 ||
+		len(input.InfrastructureProviders) > 0 ||
+		len(input.IPAMProviders) > 0 ||
+		len(input.RuntimeExtensionProviders) > 0
+
+	if isCustomUpgrade {
+		log.Logf("clusterctl upgrade apply  --core %s --bootstrap %s --control-plane %s --infrastructure %s --ipam %s --runtime-extension %s --config %s --kubeconfig %s",
+			input.CoreProvider,
+			strings.Join(input.BootstrapProviders, ","),
+			strings.Join(input.ControlPlaneProviders, ","),
+			strings.Join(input.InfrastructureProviders, ","),
+			strings.Join(input.IPAMProviders, ","),
+			strings.Join(input.RuntimeExtensionProviders, ","),
+			input.ClusterctlConfigPath,
+			input.KubeconfigPath,
+		)
+	} else {
+		log.Logf("clusterctl upgrade apply --contract %s --config %s --kubeconfig %s",
+			input.Contract,
+			input.ClusterctlConfigPath,
+			input.KubeconfigPath,
+		)
+	}
 
 	upgradeOpt := clusterctlclient.ApplyUpgradeOptions{
 		Kubeconfig: clusterctlclient.Kubeconfig{
 			Path:    input.KubeconfigPath,
 			Context: "",
 		},
-		Contract:      input.Contract,
-		WaitProviders: true,
+		Contract:                  input.Contract,
```

Comment:
Looks like we are always passing in the contract? Is that a problem for the custom upgrade?

Reply:
Looks like that case is not supported:
```go
	if ua.contract == "" && !hasProviderNames {
		return errors.New("Either the --contract flag or at least one of the following flags has to be set: --core, --bootstrap, --control-plane, --infrastructure, --ipam, --extension")
	}
	if ua.contract != "" && hasProviderNames {
		return errors.New("The --contract flag can't be used in combination with --core, --bootstrap, --control-plane, --infrastructure, --ipam, --extension")
	}
```

Would be good to handle this on all 3 places accordingly (aka only pass in what is valid + have corresponding input validation)

Reply:
Since the `ApplyUpgradeOptions` check first if `isCustomUpgrade` first else use use the contract I do not expect any problem when propagating both because we know custom providers will have priority if both are not nill 
https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/client/upgrade.go#L174-L210

Reply:
To be honest. It looks like an unintentional oversight to me that the cli is validating for this case but just the library is silently handling it in this way

@fabriziopandini WDYT? Should the lib also return an error if this invalid combination of inputs is handed over?

Reply:
and of course we check that contract is not nil if no custom provider specified before this point in the `clusterctl_helpers.go` here (although not sure how many providers needs to be checked for custom upgrade in the current case we do custom upgrade if at least one provider is not nil ) :
https://github.com/kubernetes-sigs/cluster-api/blob/aa36cbdbd020cdb254a8c170d61b6e91bc72156a/test/framework/clusterctl/clusterctl_helpers.go#L149-L157

Reply:
> @fabriziopandini WDYT? Should the lib also return an error if this invalid combination of inputs is handed over?

+1 from me, it is consistent with what we do in the CLI

Reply:
Let's continue in https://github.com/kubernetes-sigs/cluster-api/pull/7341#discussion_r992195156